### PR TITLE
fix(executor): Retry `kubectl` on transient error

### LINKED ISSUE #6467: commands.NewResourceCommand should retry when k8s api is unavailable
## Summary 

Resource executions are not retried for tcp timeouts when the K8s API becomes busy from too many workflows running in parallel. I think we can use retry.OnError from k8s.io/client-go/util/retry to avoid the issue.

```
return retry.OnError(retry.DefaultBackoff, net.IsTimeout, func() (err error) {
   // argo code here
})
```

- https://pkg.go.dev/k8s.io/client-go/util/retry#OnError
- https://pkg.go.dev/k8s.io/apimachinery/pkg/util/net#IsTimeout

## Diagnostics

👀 Yes! We need all of your diagnostics, please make sure you add it all, otherwise we'll go around in circles asking you for it:

What Kubernetes provider are you using? 

EKS 1.19

What version of Argo Workflows are you running? 

3.1.2

What executor are you running? Docker/K8SAPI/Kubelet/PNS/Emissary

n/a

Did this work in a previous version? I.e. is it a regression?

No

Logs 

```
task-ssh-config-create-update-5lwhb-1844661117: time="2021-08-02T15:28:39.528Z" level=error msg="executor error: Unable to connect to the server: dial tcp 172.20.0.1:443: i/o timeout
github.com/argoproj/argo-workflows/v3/errors.New
/go/src/github.com/argoproj/argo-workflows/errors/errors.go:49
github.com/argoproj/argo-workflows/v3/workflow/executor.(*WorkflowExecutor).ExecResource
/go/src/github.com/argoproj/argo-workflows/workflow/executor/resource.go:44
github.com/argoproj/argo-workflows/v3/cmd/argoexec/commands.execResource
/go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/resource.go:48
github.com/argoproj/argo-workflows/v3/cmd/argoexec/commands.NewResourceCommand.func1
/go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/resource.go:25
github.com/spf13/cobra.(*Command).execute
/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846
github.com/spf13/cobra.(*Command).ExecuteC
/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950
github.com/spf13/cobra.(*Command).Execute
/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main\n\t/go/src/github.com/argoproj/argo-workflows/cmd/argoexec/main.go:15\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:204
runtime.goexit
/usr/local/go/src/runtime/asm_amd64.s:1374"
```

The workflow template that timed out is creating a simple config map that stores ssh keys
```
    resource:
      action: apply
      manifest: |
        apiVersion: v1
        kind: ConfigMap
        metadata:
          name: ssh-authorized-keys
          namespace: {{inputs.parameters.envNamespace}}
          labels:
            ssh-user: {{workflow.parameters.sshUserName}}
        data:
          keys: "{{workflow.parameters.sshKeys}}"
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

### PR DESCRIPTION
**Pull request by blkperl**
Closes #6467

Signed-off-by: William Van Hevelingen <william.vanhevelingen@acquia.com>
Co-Authored-By: Glenn Pratt <glenn.pratt@acquia.com>

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

Tips:

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it does not need to pass
* Sign-off your commits to pass the DCO check: `git commit --signoff`.
* Run `make pre-commit -B` to fix codegen or lint problems. 
* Say how how you tested your changes. If you changed the UI, attach screenshots.

I'm not sure how to test this change.

### CODE REVIEW

**Review by terrytangyuan — COMMENTED** [review]
Perhaps you can test this with a TRANSIENT_ERROR_PATTERN that captures an intentional error that needs to be retried?

File: workflow/executor/resource.go [hunk 1]
```diff
@@ -34,7 +35,11 @@ func (we *WorkflowExecutor) ExecResource(action string, manifestPath string, fla
 	cmd := exec.Command("kubectl", args...)
 	log.Info(strings.Join(cmd.Args, " "))
 
-	out, err := cmd.Output()
+	var out []byte
+	err = retry.OnError(retry.DefaultBackoff, argoerr.IsTransientErr, func() error {
```
Comment (alexec) [inline workflow/executor/resource.go:39]: hmmm... what does this do, run the command X many times? 
@sarabala1979 @whynowy  I did not know about `retry.OnError` - better that argowait I think

Comment (sarabala1979) [reply workflow/executor/resource.go:39]: agree with @alexec .  `argowait` should be used.  you can enhance argowait with `onError()` if you want it.

Comment (alexec) [reply workflow/executor/resource.go:39]: sorry - I actually meant the opposite - we should stop using argowait, and use retry.OnError instead

Comment (alexec) [reply workflow/executor/resource.go:39]: Example:
```golang
	err := waitutil.Backoff(defaultRetry,
		func() (bool, error) {
			log.Infof("GCS List bucekt: %s, key: %s", artifact.GCS.Bucket, artifact.GCS.Key)
			client, err := g.newGCSClient()
			if err != nil {
				log.Warnf("Failed to create new GCS client: %v", err)
				return isTransientGCSErr(err), err
			}
			defer client.Close()
			files, err = listByPrefix(client, artifact.GCS.Bucket, artifact.GCS.Key, "")
			if err != nil {
				return isTransientGCSErr(err), err
			}
			return true, nil
		})
```
Becomes
```
	err := retry.OnError(defaultRetry, isTransientGCSErr,
		func() error {
			log.Infof("GCS List bucekt: %s, key: %s", artifact.GCS.Bucket, artifact.GCS.Key)
			client, err := g.newGCSClient()
			if err != nil {
				log.Warnf("Failed to create new GCS client: %v", err)
				return err
			}
			defer client.Close()
			files, err = listByPrefix(client, artifact.GCS.Bucket, artifact.GCS.Key, "")
			if err != nil {
				return err
			}
			return nil
		})
```

**Comment by blkperl** [comment]
It's not working yet. https://gist.github.com/blkperl/524ba2b7b0765c8d289ed3bcb892a292

It didn't detect the error as transient but it did execute the `IsTransientErr` function because `		log.Warnf("Non-transient error: %v", err)` was run

File: workflow/executor/resource.go (hunk 1 above)
Comment (sarabala1979) [reply workflow/executor/resource.go:36]: ok

**Comment by alexec** [comment]
If it is a work in progress, do you want to put it into draft?

**Comment by blkperl** [comment]
@alexec I'm not sure how to put it into a draft after the fact. I'm also still thinking about the best way to convert the error message.

**Comment by blkperl** [comment]
## 🍏 Manual Test 🍏 

### Test 1: Cluster remains unavailable
```
k logs k8s-jobs-q4x6z -f
time="2021-09-03T20:15:57.314Z" level=info msg="Starting Workflow Executor" executorType=pns version=untagged
time="2021-09-03T20:15:57.321Z" level=info msg="Creating PNS executor (namespace: default, pod: k8s-jobs-q4x6z, pid: 6)"
time="2021-09-03T20:15:57.321Z" level=info msg="Creating a K8sAPI executor"
time="2021-09-03T20:15:57.321Z" level=info msg="Executor initialized" deadline="0001-01-01 00:00:00 +0000 UTC" includeScriptOutput=false namespace=default podName=k8s-jobs-q4x6z template="{\"name\":\"main\",\"inputs\":{},\"outputs\":{},\"metadata\":{},\"resource\":{\"action\":\"apply\",\"manifest\":\"apiVersion: v1\\nkind: ConfigMap\\nmetadata:\\n  name: test\\n  namespace: default\\ndata:\\n  keys: \\\"test\\\"\\n\"}}" version="&Version{Version:untagged,BuildDate:2021-09-03T20:14:44Z,GitCommit:7858ccb90995fe33d92ddc7fdc67a6b73c63edf3,GitTag:untagged,GitTreeState:clean,GoVersion:go1.16.7,Compiler:gc,Platform:linux/amd64,}"
time="2021-09-03T20:15:57.321Z" level=info msg="Loading manifest to /tmp/manifest.yaml"
time="2021-09-03T20:15:57.330Z" level=info msg="kubectl apply -f /tmp/manifest.yaml -o json"
time="2021-09-03T20:16:27.926Z" level=info msg="Transient error: exit status 1"
time="2021-09-03T20:16:27.938Z" level=info msg="kubectl apply -f /tmp/manifest.yaml -o json"
time="2021-09-03T20:16:58.194Z" level=info msg="Transient error: exit status 1"
time="2021-09-03T20:16:58.250Z" level=info msg="kubectl apply -f /tmp/manifest.yaml -o json"
time="2021-09-03T20:17:28.389Z" level=info msg="Transient error: exit status 1"
time="2021-09-03T20:17:28.656Z" level=info msg="kubectl apply -f /tmp/manifest.yaml -o json"
time="2021-09-03T20:17:58.782Z" level=info msg="Transient error: exit status 1"
time="2021-09-03T20:17:58.782Z" level=error msg="executor error: no more retries Unable to connect to the server: dial tcp 10.152.183.1:443: i/o timeout\ngithub.com/argoproj/argo-workflows/v3/errors.Wrap\n\t/go/src/github.com/argoproj/argo-workflows/errors/errors.go:88\ngithub.com/argoproj/argo-workflows/v3/workflow/executor.(*WorkflowExecutor).ExecResource\n\t/go/src/github.com/argoproj/argo-workflows/workflow/executor/resource.go:50\ngithub.com/argoproj/argo-workflows/v3/cmd/argoexec/commands.execResource\n\t/go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/resource.go:48\ngithub.com/argoproj/argo-workflows/v3/cmd/argoexec/commands.NewResourceCommand.func1\n\t/go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/resource.go:25\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887\nmain.main\n\t/go/src/github.com/argoproj/argo-workflows/cmd/argoexec/main.go:15\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:225\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1371"
time="2021-09-03T20:17:58.783Z" level=fatal msg="no more retries Unable to connect to the server: dial tcp 10.152.183.1:443: i/o timeout\ngithub.com/argoproj/argo-workflows/v3/errors.Wrap\n\t/go/src/github.com/argoproj/argo-workflows/errors/errors.go:88\ngithub.com/argoproj/argo-workflows/v3/workflow/executor.(*WorkflowExecutor).ExecResource\n\t/go/src/github.com/argoproj/argo-workflows/workflow/executor/resource.go:50\ngithub.com/argoproj/argo-workflows/v3/cmd/argoexec/commands.execResource\n\t/go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/resource.go:48\ngithub.com/argoproj/argo-workflows/v3/cmd/argoexec/commands.NewResourceCommand.func1\n\t/go/src/github.com/argoproj/argo-workflows/cmd/argoexec/commands/resource.go:25\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887\nmain.main\n\t/go/src/github.com/argoproj/argo-workflows/cmd/argoexec/main.go:15\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:225\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1371"
```

### Test 2: Cluster recovers half way through the retries

```
k logs k8s-jobs-fhtth -f
time="2021-09-03T20:20:28.034Z" level=info msg="Starting Workflow Executor" executorType=pns version=untagged
time="2021-09-03T20:20:28.040Z" level=info msg="Creating PNS executor (namespace: default, pod: k8s-jobs-fhtth, pid: 7)"
time="2021-09-03T20:20:28.040Z" level=info msg="Creating a K8sAPI executor"
time="2021-09-03T20:20:28.041Z" level=info msg="Executor initialized" deadline="0001-01-01 00:00:00 +0000 UTC" includeScriptOutput=false namespace=default podName=k8s-jobs-fhtth template="{\"name\":\"main\",\"inputs\":{},\"outputs\":{},\"metadata\":{},\"resource\":{\"action\":\"apply\",\"manifest\":\"apiVersion: v1\\nkind: ConfigMap\\nmetadata:\\n  name: test\\n  namespace: default\\ndata:\\n  keys: \\\"test\\\"\\n\"}}" version="&Version{Version:untagged,BuildDate:2021-09-03T20:14:44Z,GitCommit:7858ccb90995fe33d92ddc7fdc67a6b73c63edf3,GitTag:untagged,GitTreeState:clean,GoVersion:go1.16.7,Compiler:gc,Platform:linux/amd64,}"
time="2021-09-03T20:20:28.041Z" level=info msg="Loading manifest to /tmp/manifest.yaml"
time="2021-09-03T20:20:28.041Z" level=info msg="kubectl apply -f /tmp/manifest.yaml -o json"
time="2021-09-03T20:20:58.159Z" level=info msg="Transient error: exit status 1"
time="2021-09-03T20:20:58.170Z" level=info msg="kubectl apply -f /tmp/manifest.yaml -o json"
time="2021-09-03T20:21:16.163Z" level=info msg="Resource: default/configmap./test. SelfLink: api/v1/namespaces/default/configmaps/test"
time="2021-09-03T20:21:16.163Z" level=info msg="No output parameters"
```

### Test 3: K8s api is stable

```
k logs k8s-jobs-2rmk8
time="2021-09-03T20:28:57.014Z" level=info msg="Starting Workflow Executor" executorType=pns version=untagged
time="2021-09-03T20:28:57.019Z" level=info msg="Creating PNS executor (namespace: default, pod: k8s-jobs-2rmk8, pid: 6)"
time="2021-09-03T20:28:57.019Z" level=info msg="Creating a K8sAPI executor"
time="2021-09-03T20:28:57.019Z" level=info msg="Executor initialized" deadline="0001-01-01 00:00:00 +0000 UTC" includeScriptOutput=false namespace=default podName=k8s-jobs-2rmk8 template="{\"name\":\"main\",\"inputs\":{},\"outputs\":{},\"metadata\":{},\"resource\":{\"action\":\"apply\",\"manifest\":\"apiVersion: v1\\nkind: ConfigMap\\nmetadata:\\n  name: test\\n  namespace: default\\ndata:\\n  keys: \\\"test\\\"\\n\"}}" version="&Version{Version:untagged,BuildDate:2021-09-03T20:14:44Z,GitCommit:7858ccb90995fe33d92ddc7fdc67a6b73c63edf3,GitTag:untagged,GitTreeState:clean,GoVersion:go1.16.7,Compiler:gc,Platform:linux/amd64,}"
time="2021-09-03T20:28:57.019Z" level=info msg="Loading manifest to /tmp/manifest.yaml"
time="2021-09-03T20:28:57.021Z" level=info msg="kubectl apply -f /tmp/manifest.yaml -o json"
time="2021-09-03T20:29:13.624Z" level=info msg="Resource: default/configmap./test. SelfLink: api/v1/namespaces/default/configmaps/test"
time="2021-09-03T20:29:13.624Z" level=info msg="No output parameters"
```

**Review by alexec — APPROVED** [review]

**Pull request merged**